src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1010 @@
+  --
+  -- *   @uniformRM (1 :: Float, 4 :: Float)@ generates values uniformly from
+  --     the set \(\{x\;|\;1 \le x \le 4\}\)
+  --
+  -- The following law should hold to make the function always defined:
+  --
+  -- > uniformRM (a, b) = uniformRM (b, a)
+  --
+  -- @since 1.2
+  uniformRM :: StatefulGen g m => (a, a) -> g -> m a
+
+instance UniformRange Integer where
+  uniformRM = uniformIntegralM
+
+instance UniformRange Natural where
+  uniformRM = uniformIntegralM
+
+instance Uniform Int8 where
+  uniformM = fmap (fromIntegral :: Word8 -> Int8) . uniformWord8
+instance UniformRange Int8 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int8 -> Word8) fromIntegral
+
+instance Uniform Int16 where
+  uniformM = fmap (fromIntegral :: Word16 -> Int16) . uniformWord16
+instance UniformRange Int16 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int16 -> Word16) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int32 where
+  uniformM = fmap (fromIntegral :: Word32 -> Int32) . uniformWord32
+instance UniformRange Int32 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int32 -> Word32) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int64 where
+  uniformM = fmap (fromIntegral :: Word64 -> Int64) . uniformWord64
+instance UniformRange Int64 where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int64 -> Word64) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Int where
+#if WORD_SIZE_IN_BITS < 64
+  uniformM = fmap (fromIntegral :: Word32 -> Int) . uniformWord32
+#else
+  uniformM = fmap (fromIntegral :: Word64 -> Int) . uniformWord64
+#endif
+  {-# INLINE uniformM #-}
+instance UniformRange Int where
+  uniformRM = signedBitmaskWithRejectionRM (fromIntegral :: Int -> Word) fromIntegral
+  {-# INLINE uniformRM #-}
+
+instance Uniform Word where
+#if WORD_SIZE_IN_BITS < 64
+  uniformM = fmap (fromIntegral :: Word32 -> Word) . uniformWord32
+#else
+  uniformM = fmap (fromIntegral :: Word64 -> Word) . uniformWord64
+#endif
+instance UniformRange Word where
+  {-# INLINE uniformRM #-}
+  uniformRM = unsignedBitmaskWithRejectionRM
+
+instance Uniform Word8 where
+  {-# INLINE uniformM #-}
+  uniformM = uniformWord8
+instance UniformRange Word8 where
+  {-# INLINE uniformRM #-}
+  uniformRM = unbiasedWordMult32RM
+
+instance Uniform Word16 where
+  {-# INLINE uniformM #-}
+  uniformM = uniformWord16
+instance UniformRange Word16 where
+  {-# INLINE uniformRM #-}
+  uniformRM = unbiasedWordMult32RM
+
+instance Uniform Word32 where
+  {-# INLINE uniformM #-}
+  uniformM  = uniformWord32
+instance UniformRange Word32 where
+  {-# INLINE uniformRM #-}
+  uniformRM = unbiasedWordMult32RM
+
+instance Uniform Word64 where
+  {-# INLINE uniformM #-}
+  uniformM  = uniformWord64
+instance UniformRange Word64 where
+  {-# INLINE uniformRM #-}
+  uniformRM = unsignedBitmaskWithRejectionRM
+
+instance Uniform CBool where
+  uniformM = fmap CBool . uniformM
```

Comment:
Let's use `Data.Coerce.coerce` here and below instead of `fmap C{Bool,Char,Int,...}` to ensure that this conversion is a no-op.

Reply:
It is not possible to use `coerce` because of `Monad m`. We could of course do `fmap coerce . uniformM`, but I don't think it'll make a difference. Let me know if I am wrong about this.

Reply:
Ah, yeah, potentially nominal roles prohibit it.